### PR TITLE
tools/ttysnoop: Introduce __ubuf_iovec support.

### PR DESCRIPTION
**Pull request by Rtoax**
kernel commit 747b1f65d39a("iov_iter: overlay struct iovec and ubuf/len") introduce 'struct iovec __ubuf_iovec' field into 'struct iov_iter', it can represent any type ITER_*, and we should use it.

Otherwise, the following code will report an error:

```
    kvec  = from->kvec;
    buf   = kvec->iov_base;
    count = kvec->iov_len;
    break;
```

On kernel versions with >=v6.3-rc4-34-g747b1f65d39a, the following error occurs:

```
    $ sudo ./ttysnoop.py 1
    ...
    ; kvec  = from->kvec;
    16: (79) r1 = *(u64 *)(r1 +16)        ; R1_w=scalar()
    ; count = kvec->iov_len;
    17: (bf) r2 = r1                      ; R1_w=scalar(id=1) R2_w=scalar(id=1)
    18: (07) r2 += 8                      ; R2_w=scalar()
    19: (05) goto pc+3
    ;
    23: (79) r2 = *(u64 *)(r2 +0)
    R2 invalid mem access 'scalar'
    processed 21 insns (limit 1000000) max_states_per_insn 0 total_states 2
    peak_states 2 mark_read 1
```

This commit solves the above error, and this submission has been tested for the following kernel versions (operating systems):

```
    Fedora 40:       6.5.0-57.fc40.x86_64
    CentOS Stream9:  5.14.0-333.el9.x86_64
    AlmaLinux 8.7:   4.18.0-425.19.2.el8_7.x86_64
```

### CODE REVIEW

File: tools/ttysnoop.py [hunk 1]
```diff
@@ -184,6 +176,37 @@ def usage():
...
+     * >= 5.14.0: case ITER_IOVEC
+     */
+    case CASE_ITER_IOVEC_NAME:
+        buf   = from->kvec->iov_base;
+        count = from->kvec->iov_len;
+        break;
+"""
+
+# kernel commit 747b1f65d39a("iov_iter: overlay struct iovec and ubuf/len")
+# we could not use field 'kvec' directly, use __ubuf_iovec instead.
```
Comment (chenhengqi) [inline tools/ttysnoop.py:191]: It seems that the commit https://github.com/torvalds/linux/commit/747b1f65d39a is compatible and is not related to the error here.

Comment (Rtoax) [reply tools/ttysnoop.py:191]: The following issues can be solved by this patch, you are right, it should be https://github.com/torvalds/linux/commit/747b1f65d39a solved the compatibility issue, not the introduction of this problem, let me modify the following comment.

```bash
$ sudo ./ttysnoop.py 3
...
bpf: Failed to load program: Permission denied
reg type unsupported for arg#0 function kfunc__vmlinux__tty_write#11
0: R1=ctx(off=0,imm=0) R10=fp0
; KFUNC_PROBE(tty_write, struct kiocb *iocb, struct iov_iter *from)
0: (79) r2 = *(u64 *)(r1 +0)
func 'tty_write' arg0 has btf_id 830 type STRUCT 'kiocb'
1: R1=ctx(off=0,imm=0) R2_w=ptr_kiocb(off=0,imm=0)
; if (iocb->ki_filp->f_inode->i_ino != 6)
1: (79) r2 = *(u64 *)(r2 +0)          ; R2_w=ptr_file(off=0,imm=0)
; if (iocb->ki_filp->f_inode->i_ino != 6)
2: (79) r2 = *(u64 *)(r2 +168)        ; R2_w=ptr_inode(off=0,imm=0)
; if (iocb->ki_filp->f_inode->i_ino != 6)
3: (79) r2 = *(u64 *)(r2 +64)         ; R2_w=scalar()
; if (iocb->ki_filp->f_inode->i_ino != 6)
4: (55) if r2 != 0x6 goto pc+396      ; R2_w=6
; KFUNC_PROBE(tty_write, struct kiocb *iocb, struct iov_iter *from)
5: (79) r1 = *(u64 *)(r1 +8)
func 'tty_write' arg1 has btf_id 932 type STRUCT 'iov_iter'
6: R1_w=ptr_iov_iter(off=0,imm=0)
; if (from->iter_type != ITER_UBUF && from->iter_type != ITER_IOVEC)
6: (71) r2 = *(u8 *)(r1 +0)           ; R1_w=ptr_iov_iter(off=0,imm=0) R2_w=scalar(umax=255,var_off=(0x0; 0xff))
; if (from->iter_type != ITER_UBUF && from->iter_type != ITER_IOVEC)
7: (15) if r2 == 0x5 goto pc+1        ; R2_w=scalar(umax=255,var_off=(0x0; 0xff))
8: (55) if r2 != 0x0 goto pc+392      ; R2=0
; if (from->data_source != WRITE)
9: (71) r3 = *(u8 *)(r1 +3)           ; R1=ptr_iov_iter(off=0,imm=0) R3_w=scalar(umax=255,var_off=(0x0; 0xff))
; if (from->data_source != WRITE)
10: (15) if r3 == 0x0 goto pc+390     ; R3_w=scalar(umax=255,var_off=(0x0; 0xff))
; switch (from->iter_type) {
11: (15) if r2 == 0x5 goto pc+8       ; R2=0
12: (b7) r7 = 0                       ; R7_w=0
13: (b7) r3 = 0                       ; R3_w=0
14: (7b) *(u64 *)(r10 -16) = r3       ; R3_w=0 R10=fp0 fp-16_w=00000000
15: (55) if r2 != 0x0 goto pc+10      ; R2=0
; kvec  = from->kvec;
16: (79) r1 = *(u64 *)(r1 +16)        ; R1_w=scalar()
; count = kvec->iov_len;
17: (bf) r2 = r1                      ; R1_w=scalar(id=1) R2_w=scalar(id=1)
18: (07) r2 += 8                      ; R2_w=scalar()
19: (05) goto pc+3
;
23: (79) r2 = *(u64 *)(r2 +0)
R2 invalid mem access 'scalar'
processed 21 insns (limit 1000000) max_states_per_insn 0 total_states 2 peak_states 2 mark_read 1

Traceback (most recent call last):
  File "/home/sdb/Git/bcc/tools/./ttysnoop.py", line 236, in <module>
    b = BPF(text=bpf_text)
        ^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/bcc/__init__.py", line 487, in __init__
    self._trace_autoload()
  File "/usr/lib/python3.12/site-packages/bcc/__init__.py", line 1483, in _trace_autoload
    self.attach_kfunc(fn_name=func_name)
  File "/usr/lib/python3.12/site-packages/bcc/__init__.py", line 1145, in attach_kfunc
    fn = self.load_func(fn_name, BPF.TRACING)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/bcc/__init__.py", line 526, in load_func
    raise Exception("Failed to load BPF program %s: %s" %
Exception: Failed to load BPF program b'kfunc__vmlinux__tty_write': Permission denied
```

Comment (Rtoax) [reply tools/ttysnoop.py]: Of course, this patch can also solve the above problem, however, it is not yet known what the root cause is

**Comment by Rtoax** [comment]
The problem of this PR try to fix is solved by commit https://github.com/iovisor/bcc/commit/ce5c8938c494eb03f0784c6e4ae81507139ca779, I just close this PR.

**Pull request closed without merging**